procrustes/permutation.py

```diff
@@ -371,25 +370,32 @@ def permutation_2sided(array_a, array_b, transform_mode="single",
 
     """
     # check inputs
-    new_a, new_b = setup_input_arrays(array_a, array_b, remove_zero_col, remove_zero_row,
-                                      pad_mode, translate, scale, check_finite, weight)
-    # np.power() can not handle the negatives values
-    # Try to convert the matrices to non-negative
-    # shift the the matrices to avoid negative values
-    # otherwise it will cause an error in the Eq. 28 in the research notes
-    maximum = max(np.amax(np.abs(new_a)), np.amax(np.abs(new_b)))
-    new_a_positive = new_a.astype(np.float) + maximum
-    new_b_positive = new_b.astype(np.float) + maximum
+    new_a, new_b = setup_input_arrays(array_a, array_b, unpad_col, unpad_row,
+                                      pad, translate, scale, check_finite, weight)
+    if transform_mode == "single":
+        if new_a.shape != new_b.shape:
+            raise ValueError(
+                f"Shape of A and B does not match: {new_a.shape} != {new_b.shape} "
+                "Check pad, remove_zero_col, and remove_zero_row arguments."
+            )
+
     # Do single-transformation computation if requested
     transform_mode = transform_mode.lower()
     if transform_mode == "single":
+        # np.power() can not handle the negatives values
+        # Try to convert the matrices to non-negative
+        # shift the the matrices to avoid negative values
+        # otherwise it will cause an error in the Eq. 28 in the research notes
+        maximum = max(np.amax(np.abs(new_a)), np.amax(np.abs(new_b)))
+        new_a_positive = new_a.astype(np.float) + maximum
+        new_b_positive = new_b.astype(np.float) + maximum
+
```

Comment:
I'm not sure why this part is needed to make sure that it is positive.  The comments aren't obvious to me why np.power can't handle negative values or what Eq. 28 is (maybe it is Eq B.29?). I was wondering if @fwmeng88 remembers?  If it is Eq B.29, it seems that it should be able to handle negative values.

Reply:
This comment corresponds to the Equation 28 in the notes. I will email you the notes shortly. The reason why we need to ensure the non-negativity is square root does not allow negative values. Eq 28 can be generalized to Eq. 8 and Eq. 14 in paper [Nonnegative matrix factorization for combinatorial optimization: Spectral clustering, graph matching, and clique finding](https://ieeexplore.ieee.org/abstract/document/4781113/?casa_token=GNr6vDvB6NoAAAAA:Jks0I4SEV8FG9Exn-5qL-wtCReYTmCvZBIXtshjC7wAwmQFR4WpO4iw-_y4Z95-kfo6r90Qeow), which is exactly what we implemented for directed and undirected.

If we comment on these lines of code, some tests will fail. Hope this helps a little. @Ali-Tehrani

Reply:
Thank you it's a lot more clear now.

Reply:
power can handle square rooting of negative values, it just results in a complex number that we are not covering at this stage.
Also, please clarify the comments as just mentioning Eq. 28 is not helpful.